### ISSUE TITLE
[a11y]: Modal's `launcherButtonRef` does not work with `enable-dialog-element` flag

### ISSUE THREAD

**Issue by wkeese** [carbon-design-system/carbon#19404]
### Package

@carbon/react

### Browser

Chrome

### Operating System

MacOS

### Package version

1.82

### React version

18

### Automated testing tool and ruleset

N/A

### Assistive technology

none

### Description

This is about Modals when using the `enableDialogElement` feature flag.

In some situations, when you close the Modal, focus goes back to `<body>` rather than where the modal was launched from.

This happens even when using `launcherButtonRef`.

So there are two issues:

1. Poor accessibility.  User needs to tabs dozens of times to get back to where they were before.
2. Modal not working according to spec.

Specifically, the spec implies that `launcherButtonRef` should work regardless of any feature flag settings:

```js
/**
 * Provide a ref to return focus to once the modal is closed.
 */
launcherButtonRef?: Ref<HTMLButtonElement>;
```

(I interpret that description to mean that Modal must focus `launcherButtonRef` on close, regardless of what node `launcherButtonRef` points to.)

### Reproduction/example

https://stackblitz.com/edit/github-ykbrpwcg-eyfjd2nc?file=src%2FExample%2FExample.jsx

### Steps to reproduce

Try launching a modal from the button on https://stackblitz.com/edit/github-ykbrpwcg-eyfjd2nc?file=src%2FExample%2FExample.jsx.  When you close it, focus returns to the launch button, as expected.  This is thanks to the native support of the `<dialog>` component to do the right thing.

However, try launching a modal from the overflow menu.  When you close it, the focus goes to `<body>`, even though I specifically set `launcherButtonRef` so that wouldn't happen.  This is not good.

The workaround is to set focus on the `onRequestClose` and `onRequestSubmit` handlers, probably wrapped in a `setTimeout()`.   But I think that (as stated in the documentation), Modal should obey the setting of `launcherButtonRef`.

### Suggested Severity

None

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [x] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

**Comment by wkeese** [comment]
> Would we be able to "wait" for the browser to do it's native return-focus behavior and then focus launcherButtonRef if the browser focused the body?

This is basically what I expected, although I would phrase it as:

> If the caller specifies `launcherButtonRef`, it overrides the browser's native focus behavior.

It seems trivial to implement by just waiting for browser to do its native focus first, and then refocusing `launcherButtonRef` afterwards.

The current code already has a `setTimeout()`, so wouldn't this happen automatically?   Just remove the `enableDialogElement` test from:

```js
useEffect(() => {
  if (!enableDialogElement && prevOpen && !open && launcherButtonRef) {
    setTimeout(() => {
      if ('current' in launcherButtonRef) {
        launcherButtonRef.current?.focus();
      }
    });
  }
}, [open, prevOpen, launcherButtonRef, enableDialogElement]);
```

So it becomes:

```js
useEffect(() => {
  if (prevOpen && !open && launcherButtonRef) {
    setTimeout(() => {
      if ('current' in launcherButtonRef) {
        launcherButtonRef.current?.focus();
      }
    });
  }
}, [open, prevOpen, launcherButtonRef]);
```

Did I miss something?

I'm assuming that Carbon V12 will drop the 'enable-experimental-focus-wrap-without-sentinels' and  'enable-dialog-element' feature flags, and standardize on using `<dialog>`.  In that case, continuing to support `launcherButtonRef` would be the easiest upgrade path.

**Comment by tay1orjones** [comment]
It looks like the close event isn't async, https://github.com/whatwg/html/issues/5678#issuecomment-832288148, so yeah your suggestion would work I think. We'd need to test across browsers, but I assume they've all got this implemented correctly to where we could rely on these being fired in order. https://github.com/whatwg/html/issues/6681

We're not totally sure yet on if or how Dialog will replace Modal, but I like the idea of naming this `focusAfterCloseRef` on `Dialog`. Modal could map `launcherButtonRef` to it internally. If at some point in the future it's decided that Modal usages should migrate to using Dialog directly, we can smooth out that upgrade path. Support `launcherButtonRef` on `Dialog` with a deprecation/console warning, codemod the Modal->Dialog migration and handle the prop name change there, etc.
